### PR TITLE
Cancel HNSW search requests that have already timed out from a user's POV

### PR DESCRIPTION
**Pull request by etiennedi**
### What's being changed:
A user-initiated request will eventually time out. This could be because the user set a client-side timeout or because a server-side timeout kicked in. However, if the search had not completed yet, it would keep running "forever". This fixes this by making sure the search methods respect the passed-in context.

Note that this does not address the root cause of a search taking so long it would time out. This could be because of the system being overloaded or other resource constraints. However, in practice, we mainly see that with extremely selective filters. A much better alternative for those scenarios is present in v1.27+ with ACORN-like search.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

**Comment by etiennedi** [comment]
I did the following simple experiment to validate it's working.

I added a log line in each candidate iteration in HNSW. You can see that a normal query takes about 512 iterations and takes 10ms (with the additional logging):

Screenshot 2024-10-29 at 11 24 03 AM

Then I added a 1s delay every time we print this message. The client has a 20s timeout configured. After ~20s the client times out and the query is indeed canceled correctly:

Screenshot 2024-10-29 at 11 25 07 AM

_Note: The reason it stops after 18s and not 20s is because the method is called multiple times (once per layer), but I'm only measuring the time per layer, not the total time._

**Comment by etiennedi** [comment]
I was wondering if this had any effect on the query slow log, so I verified that the slow log still works – despite the cancellation. It does:

image

File: adapters/repos/db/vector/hnsw/delete.go [hunk 1]
```diff
@@ -407,13 +411,15 @@ func tombstoneDeletionConcurrency() int {
...
+func (h *hnsw) reassignNeighborsOf(ctx context.Context, deleteList helpers.AllowList,
+	breakCleanUpTombstonedNodes breakCleanUpTombstonedNodesFunc,
+) (ok bool, err error) {
 	h.RLock()
 	size := len(h.nodes)
 	h.RUnlock()
 
-	g, ctx := enterrors.NewErrorGroupWithContextWrapper(h.logger, h.shutdownCtx)
-	ctx, cancel := context.WithCancel(ctx)
+	g, shutdownContext := enterrors.NewErrorGroupWithContextWrapper(h.logger, h.shutdownCtx)
```
Comment (moogacs) [inline adapters/repos/db/vector/hnsw/delete.go:421]: **question:** double check,  is that meant to be `h.shutdownCtx` or `ctx` ?

Comment (reyreaud-l) [reply adapters/repos/db/vector/hnsw/delete.go:421]: Agree, also it doesn't look like we're using the `ctx` we're passed as parameter. Is `h.shutdownCtx` different ?

Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/delete.go:421]: It was previously called `ctx` and I renamed the existing `ctx` to `shutdownContext` as that what it' seems to be doing. Then renaming all previous uses. The new usage (`ctx` passed in as an arg) is used in line 450.

There is a good chance we could use the same context, yeah, but I'm not really sure how to verify this.

File: adapters/repos/db/vector/hnsw/search.go [hunk 2]
```diff
@@ -210,6 +216,13 @@ func (h *hnsw) searchLayerByVectorWithDistancer(queryVector []float32,
 	connectionsReusable := make([]uint64, h.maximumConnectionsLayerZero)
 
 	for candidates.Len() > 0 {
+		time.Sleep(1 * time.Second)
```
Comment (reyreaud-l) [inline adapters/repos/db/vector/hnsw/search.go:219]: This adds an effective one second sleep in what looks like the core loop of the search, are we ok with that ?

Cancellation looks OK to me :)

Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/search.go]: Thanks! That was of course not intentional. The last commit added this (it was local change for testing). Removed again!

File: adapters/repos/db/vector/hnsw/delete.go (hunk 1 above)
Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/delete.go]: Pushed a new commit that should simplify things.

**Review by reyreaud-l — APPROVED** [review]

**Pull request merged**